interface/modules/custom_modules/oe-module-faxsms/src/Controller/AppDispatch.php

```diff
@@ -49,6 +50,8 @@ public function __construct($type = null)
         if (empty(self::$_apiModule)) {
             self::$_apiModule = $_REQUEST['type'] ?? $_SESSION["oefax_current_module_type"] ?? null;
         }
+        $this->crypto = new CryptoGen();
+        $this->getCredentials();
```

Comment:
Does this make sense?  It seems like for API handling we were having to sprinkle calls to `getCredentials()` around to make sure that the config had been loaded (and triggered the corresponding side effects).  This front loads it so that it's ready before calling `dispatchAction()`.

Reply:
What side effects? I prefer to call credentials from client on demand. Please leave this way.

Reply:
I had originally seen/modeled on the Twilio integration, which [sets fields](https://github.com/openemr/openemr/blob/master/interface/modules/custom_modules/oe-module-faxsms/src/Controller/TwilioSMSClient.php#L76-L82) as a side effect of loading the credentials.  The [email](https://github.com/openemr/openemr/blob/master/interface/modules/custom_modules/oe-module-faxsms/src/Controller/EmailClient.php#L51-L56) and [EtherFAX](https://github.com/openemr/openemr/blob/master/interface/modules/custom_modules/oe-module-faxsms/src/Controller/EtherFaxActions.php#L65-L69) integrations do the same thing, though the RingCentral one [does not](https://github.com/openemr/openemr/blob/master/interface/modules/custom_modules/oe-module-faxsms/src/Controller/RCFaxClient.php#L236).

Looking more closely, many of those seem like they were just copy/paste and not needed.  Also, all but email *also* call `getCredentials()` in the subclass constructor, and then store that, as well.

Given all of that, I'd lean towards adding a protected `credentials` field to `AppDispatch`, and populating it by calling `getCredentials()` in `AppDispatch::_construct()`.  That may mean cleaning up all the others that have already declared `credentials` fields - I'm not a PHP dev so not sure how the language handles field hiding.  (Then the side effects of setting fields in `getCredentials()` could also be removed.)

How would you like me to handle this?

Reply:
Again like I said I had this in the beginning and had to change because of an issue I can't remember. Could have been contention or scope was lost.
You are stating the more logical way but I'm sure I did this by design. Why are you having a problem calling from your client?
Also RC used to be 2FA with caching tokens so was handled differently until I recently went JWT.

I tell you what. I have some refactors to do for our next release so I'll clean this issue up then when I can do a full test suite.
I'm just fearful of any patches until next release having this change done.

Reply:
> Why are you having a problem calling from your client?

I could certainly make a late call to `getCredentials()` in `sendSMS()`.  I've been trying to take an eye to "how can the next integration be made simpler" as I've run into things.  I'm quite happy to whichever approach you'd prefer; these are just my takes on potential simplifications.

This particular one (credentials) is interesting because of the timing of `dispatchAction()` - it's invoked at the end of `AppDispatch::__construct()`.  That means that e.g. `sendSms()` will be called _before_ `parent::__construct()` returns.  I believe that this is why `crypto` was being populated in the sub classes `__construct()` - they also had to call `getCredentials()` (which depends on `crypto`) *before* invoking `parent::__construct()` to have them before `sendSMS()` was called, and so had to make sure that `crypto` was populated.

One alternative is to JIT load credentials where/when they are needed.  This has the downside of having potentially multiple locations in an integration that do so.  Another alternative is to populate a `credentials` field once in `AppDispatch::__construct()` so that it will be available always.

(I will note that the current approach of copying data to properties from credentials has the benefit of localizing the extraction/mapping of credential fields to named fields to a single place, rather than potentially multiple places that credentials are used.  That can't be done without using side effects due to the timing of the `parent::__construct()` invocation.)

I understand your concern about keeping changes minimal for a patch.  What's your preferred approach:
1. Keep current implementation of calling `getCredentials()` in `AppDispatch::__construct()` and use side effects to capture what's needed (this means that `getCredentials()` will be called twice for most current integrations)
2. Instantiate `crypto` in `ClickatellSMSClient::__construct()` and call `getCredentials()` directly (like other integrations)
3. Wait and call `getCredentials()` in `sendSMS()`, since for this integration that's the only place they will be needed

I'll happily take whichever approach you'd like.

Reply:
Okay as long as we don't have static properties for credentials which I don't think we have then should be okay.
Still for some reason I remember an issue concerning background task requests and credentials so maybe I'll take the time to research just to satisfy myself. Keep as you have it. I may go ahead and refactor other clients and test.